File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/aws/dynamodb/DynamoDBUtils.java
package uk.nhs.nhsx.core.aws.dynamodb;

import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClientBuilder;
import com.amazonaws.services.dynamodbv2.document.*;
import com.amazonaws.services.dynamodbv2.document.spec.DeleteItemSpec;
import com.amazonaws.services.dynamodbv2.model.ReturnValue;

public class DynamoDBUtils implements AwsDynamoClient {

	private final DynamoDB dynamoDB;

	public DynamoDBUtils() {
		dynamoDB = new DynamoDB(AmazonDynamoDBClientBuilder.defaultClient());
	}

	@Override
	public void putItem (String tableName, Item item) {
		Table table = dynamoDB.getTable(tableName);
    	table.putItem(item);
	}

	@Override
	public Item getItem(String tableName, String hashKeyName, String hashKeyValue) {
		Table table = dynamoDB.getTable(tableName);
		return table.getItem(hashKeyName, hashKeyValue);
	}

	@Override
	public DeleteItemOutcome deleteItem(String tableName, String hashKeyName, String hashKeyValue) {
		Table table = dynamoDB.getTable(tableName);

		DeleteItemSpec deleteItemSpec = new DeleteItemSpec()
			.withPrimaryKey(new PrimaryKey(hashKeyName, hashKeyValue))
			.withReturnValues(ReturnValue.ALL_OLD);

		return table.deleteItem(deleteItemSpec);
	}
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/aws/dynamodb/DynamoDBUtils.kt
package uk.nhs.nhsx.core.aws.dynamodb

import com.amazonaws.services.dynamodbv2.AmazonDynamoDB
import com.amazonaws.services.dynamodbv2.document.DeleteItemOutcome
import com.amazonaws.services.dynamodbv2.document.DynamoDB
import com.amazonaws.services.dynamodbv2.document.Item
import com.amazonaws.services.dynamodbv2.document.spec.DeleteItemSpec
import com.amazonaws.services.dynamodbv2.document.PrimaryKey
import com.amazonaws.services.dynamodbv2.model.ReturnValue

class DynamoDBUtils(client: AmazonDynamoDB) : AwsDynamoClient {

    private val dynamoDB: DynamoDB = DynamoDB(client)

    override fun getItem(
        tableName: String,
        hashKeyName: String,
        hashKeyValue: String
    ): Item? = dynamoDB.getTable(tableName).getItem(hashKeyName, hashKeyValue)

    override fun deleteItem(
        tableName: String,
        hashKeyName: String,
        hashKeyValue: String
    ) {
        dynamoDB.getTable(tableName).deleteItem(
            DeleteItemSpec()
                .withPrimaryKey(PrimaryKey(hashKeyName, hashKeyValue))
                .withReturnValues(ReturnValue.ALL_OLD)
        )
    }
}
